File: src/algorithms/arrays/two-pointer/container-with-most-water/sources/container-with-most-water.py
# Container With Most Water — two pointers converge inward, always moving the shorter bar to maximize area
def container_with_most_water(heights: list[int]) -> dict:
    left_pointer = 0  # @step:initialize
    right_pointer = len(heights) - 1  # @step:initialize
    max_area = 0  # @step:initialize
    best_left = 0  # @step:initialize
    best_right = len(heights) - 1  # @step:initialize

    while left_pointer < right_pointer:
        left_height = heights[left_pointer]  # @step:visit
        right_height = heights[right_pointer]  # @step:visit
        current_area = min(left_height, right_height) * (right_pointer - left_pointer)  # @step:compare

        if current_area > max_area:  # @step:compare
            max_area = current_area  # @step:compare
            best_left = left_pointer  # @step:compare
            best_right = right_pointer  # @step:compare

        if left_height <= right_height:  # @step:compare
            left_pointer += 1  # @step:visit
        else:
            right_pointer -= 1  # @step:visit

    return {"max_area": max_area, "left_index": best_left, "right_index": best_right}  # @step:complete

File: src/algorithms/arrays/two-pointer/container-with-most-water/sources/container-with-most-water.go
// Container With Most Water — two pointers converge inward, always moving the shorter bar to maximize area
package containerwithmostwater

func containerWithMostWater(heights []int) (maxArea int, leftIndex int, rightIndex int) {
	leftPointer := 0               // @step:initialize
	rightPointer := len(heights) - 1 // @step:initialize
	maxArea = 0                    // @step:initialize
	bestLeft := 0                  // @step:initialize
	bestRight := len(heights) - 1  // @step:initialize

	for leftPointer < rightPointer {
		leftHeight := heights[leftPointer]   // @step:visit
		rightHeight := heights[rightPointer] // @step:visit
		minHeight := leftHeight
		if rightHeight < minHeight {
			minHeight = rightHeight
		}
		currentArea := minHeight * (rightPointer - leftPointer) // @step:compare

		if currentArea > maxArea { // @step:compare
			maxArea = currentArea    // @step:compare
			bestLeft = leftPointer   // @step:compare
			bestRight = rightPointer // @step:compare
		}

		if leftHeight <= rightHeight { // @step:compare
			leftPointer++ // @step:visit
		} else {
			rightPointer-- // @step:visit
		}
	}

	return maxArea, bestLeft, bestRight // @step:complete
}
